src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+            {
+                var parentStatement = invocationExpression.GetAncestorOrThis<StatementSyntax>();
+                if (parentStatement != null)
+                {
+                    if (TryConvertIfInInvocationInternal(
+                            source,
+                            invocationExpression,
+                            parentStatement,
+                            initializer,
+                            variableName,
+                            out var variable,
+                            out var documentCreationMethod))
+                    {
+                        if (TryConvertQueryExpression(
+                            source,
+                            expression => leafExpressionCreationMethod(variable, expression),
+                            out var statement))
+                        {
+                            documentUpdateInfo = documentCreationMethod(statement);
+                            return true;
+                        }
+                    }
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInInvocationInternal(
+                QueryExpressionSyntax source,
+                InvocationExpressionSyntax invocationExpression,
+                StatementSyntax parentStatement,
+                ExpressionSyntax initializer,
+                string variableName,
+                out ExpressionSyntax variable,
+                out Func<StatementSyntax, DocumentUpdateInfo> documentUpdateMethod)
+            {
+                var invocationParent = invocationExpression.Parent;
+                switch (invocationParent.Kind())
+                {
+                    case SyntaxKind.EqualsValueClause:
+                        if (invocationParent.Parent is VariableDeclaratorSyntax variableDeclarator)
+                        {
+                            // before var a = (from a in b select a).ToList();
+                            // after var a = new List<T>();
+                            // foreach(...)
+                            variable = SyntaxFactory.IdentifierName(variableDeclarator.Identifier);
+                            documentUpdateMethod =
+                                s => new DocumentUpdateInfo(
+                                    parentStatement,
+                                    new[] { parentStatement.ReplaceNode(invocationExpression, initializer.WithAdditionalAnnotations(Simplifier.Annotation)), s });
+                            return true;
+                        }
+                        break;
+                    case SyntaxKind.SimpleAssignmentExpression:
+                        // before a = (from a in b select a).ToList();
+                        // after a = new List<T>();
+                        // foreach(...)
+                        variable = ((AssignmentExpressionSyntax)invocationParent).Left;
```

Comment:
>variable = ((AssignmentExpressionSyntax)invocationParent).Left; [](start = 24, length = 63)

The same question for types. Even if types align, the expression might be a right only property, or it could have a bunch of side-effects that shouldn't be evaluated multiple times in the loop. Even if it is a simple read-right property of the right type, where is a guarantee that its getter returns a value that was previously passed to the setter?